### ISSUE TITLE
cssnano is slow (takes 16.6s for 334KB CSS file)

### ISSUE THREAD

**Issue by paulschreiber** [ben-eb/cssnano#47]
cssnano is slow. It takes 16.6s for my 334KB CSS file.

cssnano 2.3.0
OS X 10.10.5 (14F27)
rMBP / 2.3 GHz Core i7 / 16 GB RAM

```
$ time cssnano < assets/css/fivethirtyeight.src.css > /tmp/x

real    0m16.618s
user    0m16.561s
sys 0m0.131s
```

To compare, ycssmin takes 0.16s:

```
$ time ycssmin  < assets/css/fivethirtyeight.src.css > /tmp/y

real    0m0.162s
user    0m0.137s
sys 0m0.026s
```

Here's the sample CSS:
https://gist.github.com/paulschreiber/efbe30ac4d693296c299

**Comment by ben-eb** [comment]
16 seconds is a lot, I've not seen anything like that kind of slowness before. For the record I run a similar setup to you, 2.5Ghz i7 with 16GB of RAM (mid-range MBP) and I get this:

```
real	0m5.618s
user	0m5.706s
sys	0m0.056s
```

Speed is one of the areas in which we are trying to improve, but as you might appreciate, one of the downsides of modularity means that work gets performed repetitively. This was by design, as it is a lot easier to write a compression tool that has little parts that each do one thing well. However it does also mean that performance is not as fast as some other tools.

But comparing to ycssmin is not really fair, as cssnano does many complex operations not typically seen from older minifiers. For example, ycssmin incorrectly compresses `h1{clip: rect(0px,0px,0px,0px)}` to `h1{clip:rect(0)}`. cssnano instead parses the CSS file and identifies areas which can be compressed, so naturally it will be slower than naively searching and replacing the CSS for certain values.

**Comment by ai** [comment]
@paulschreiber what env to you use to run cssnano? node.js 0.10?

**Comment by paulschreiber** [comment]
At home (where it's taking 14.1s) I'm using these versions of node and npm:

$ npm -v
2.13.4
$ node -v
v0.12.0

I ran `sample` on node:
https://gist.githubusercontent.com/paulschreiber/49623d9a6bba3ef056fa/raw/679b50483fbd8fbf7b10b08915ba48b70d8b4247/node_2015-08-17_225043_5GlG.sample.txt

Unfortunately, I can't make too much sense of that. It's all in v8 code and lacks symbols.

**Comment by paulschreiber** [comment]
Another comparison:
```
$ time cleancss  ~/.dev/espn-fivethirtyeight/assets/css/fivethirtyeight.src.css > /tmp/x

real	0m0.693s
user	0m0.694s
sys	0m0.054s
```

Something in this file is causing degenerate behaviour in cssnano.

**Comment by paulschreiber** [comment]
Even with a one-line file, cssnano has a lot of startup overhead / is slower. It was 0.45s on first run.

```
$ time echo ".newsletter-checkbox { position: relative; }" > /dev/null | cssnano
real	0m0.340s
user	0m0.317s
sys	0m0.040s

$ time echo ".newsletter-checkbox { position: relative; }" > /dev/null | cleancss
real	0m0.165s
user	0m0.147s
sys	0m0.023s

$ time echo ".newsletter-checkbox { position: relative; }" > /dev/null | ycssmin
real	0m0.088s
user	0m0.072s
sys	0m0.016s
```

**Comment by denji** [comment]
http://goalsmashers.github.io/css-minification-benchmark/

**Comment by infobip-team-webs** [comment]
After some testing, seems that [discardDuplicates](https://github.com/ben-eb/postcss-discard-duplicates) plugin is slowest in the stack. When I set `discardDuplicates: false`  my build time is ~3s, with plugin turned on it's ~37s.

It would be worth looking into [`dedupe`](https://github.com/ben-eb/postcss-discard-duplicates/blob/master/index.js) for possible optimizations.

**Comment by teodragovic** [comment]
After some testing, seems like [discardDuplicates](https://github.com/ben-eb/postcss-discard-duplicates) is the slowest in the stack. With `discardDuplicates: false` setting, my build time is ~3s. With plugin turned on, it's ~37s.

It would be worth checking [`dedupe`](https://github.com/ben-eb/postcss-discard-duplicates/blob/master/index.js) function for possible optimizations.

**Comment by ben-eb** [comment]
@teodragovic Thanks for the report, had a look into this today. We made some optimisations that result in a ~50% speed increase, so now I get this time when processing the OP's CSS:

```
real	0m2.725s
user	0m2.792s
sys	0m0.072s
```

https://github.com/ben-eb/postcss-discard-duplicates/commit/286e17cc2fa9bce2bbcb5695bd6f2ba3d923d52d

**Comment by teodragovic** [comment]
Awesome!

I did quick run and my compile time is now at ~13.5 sec. Also managed to shave one second by replacing `forEach()` with regular for loop and `String()` with `toString()`.

**Comment by ben-eb** [comment]
@teodragovic Something like this?

```js
function dedupe (node, index) {
    if (node.type === 'comment') { return; }
    if (node.nodes) { node.each(dedupe); }

    var toString = node.toString();
    var nodes = node.parent.nodes;
    var result = [node];
    var i = index + 1;
    var max = nodes.length;

    for (; i < max; i++) {
        if (nodes[i].toString() === toString) {
            result.push(nodes[i]);
        }
    }

    for (var j = 0, jmax = result.length - 1; j < jmax; j++) {
        result[j].removeSelf();
    }
}
```

**Comment by teodragovic** [comment]
@ben-eb (sorry, I didn't get gh notification for new post)

I used slightly different for loop but this one works the same as far I could tell. Replacing `removeSelf()` with `remove()` would shave off additional ~300ms.

**Comment by TrySound** [comment]
@teodragovic This is old API issue. We will move cssnano after autoprefixer release.

**Comment by ben-eb** [comment]
Performance should be improved further with the latest version of PostCSS, as we use `each` a lot.

https://github.com/postcss/postcss/commit/d8ecd9ada4298cbdca350a941d3bf1a8fcd6eecc

**Comment by mdaomega** [comment]
I have the same problem. see this link:

https://github.com/laravel/elixir/issues/460

tests have been run minutes ago. note the post that i said: 

> it's funny. just by adding cssnano = require( "gulp-cssnano" ), to gulpfile, without applying .pipe( cssnano() ) running gulp delays 4 seconds

**Comment by sondreb** [comment]
I just want to give my comment that gulp-cssnano (2.1.2) is very slow compared to gulp-clean-css (2.0.10). cssnano uses **1-1.5 minute** to minimize our library css (concats of third party libaries) which is 1MB original and ends up as 742 minimized.

clean-css takes **3 seconds** on the same file, 755KB result.

Both libraries are used without any options.

I have no option to stop using cssnano and I hope you can improve this in the future.

**Comment by denji** [comment]
@sondreb https://github.com/lahmatiy/postcss-csso

**Comment by nalbion** [comment]
it was taking 22 seconds for me.  As per the [Frank N's suggestion on SO](http://stackoverflow.com/a/35598145/1225993) I've switched to clean-css, which seems to produce [better results](http://goalsmashers.github.io/css-minification-benchmark/) also.

**Comment by ben-eb** [comment]
I'm going to mark this as resolved; postcss-discard-duplicates 2.0.2 has a much improved algorithm which is much less complex. New installs should see a drop in build times especially for large CSS files.
